### PR TITLE
[cmake] Remove broken check from ROOT_GENERATE_DICTIONARY

### PR DESCRIPTION
**Pull request by hahnjo**
Recently, I am seeing test failures in local builds depending on the build directory, or rather the length of its name / path. I bisected this to commit https://github.com/root-project/root/commit/46ba2eefd00371b024ae239626c7600cd3259885 ("Enable a few modules for Windows") which introduced a check to filter dependencies without cxxmodule.

There are (at least) two problems with this check: First, we add the `cpp_module` which is equal to the `library_target_name` and not the pcm file name. This results in currently no module dependencies being passed to `rootcling` at all because the condition is never true!

This could be fixed by checking `dep IN_LIST local_modules_idx_deps`. However, that still doesn't work because `local_modules_idx_deps` does not contain all modules from the start but is constructed on-the-fly. This is particularly a problem for `libCore` whose dictionary is only created after the other libraries located below `core/` are processed, for example `Imt`. The problem is that `libImt` depends on `Core`, but at the time `ROOT_GENERATE_DICTIONARY` is called for `Imt`, it doesn't know yet about the dictionary of `Core`.

Finally, I'm not sure if completely omitting the dependency for a target with `NO_CXXMODULE` in case of `runtime_cxxmodules` is correct. Based on the code around, I would expect to fall back to `*_rdict.pcm` in that case.

Note: This change will probably break the (partial) support for modules on Windows, but for the the reasons spelled out above, I don't see an easy fix given the current situation.

### CODE REVIEW

**Comment by hahnjo** [comment]
I'm marking this as critical because I don't know the implications of not specifying module dependencies to `rootcling`. It seems to break some cases for me (even though I don't understand what the length of the build directory name has to do with it...), even though all CI setups seem to work (by chance?)

File: roofit/roofitcore/CMakeLists.txt [hunk 1]
```diff
@@ -476,9 +476,9 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooFitCore
   DICTIONARY_OPTIONS
     "-writeEmptyRootPCM"
   LIBRARIES
+    RooBatchCompute
```
Comment (Axel-Naumann) [inline roofit/roofitcore/CMakeLists.txt:479]: Doesn't think link `RooBatchCompute`? I thought the purpose was to `dlopen` it as needed?

**Comment by vgvassilev** [comment]
I am a little confused on how that affects path length. 

The idea of that patch was that if we specify no_cxxmodule we should not expect targets that depend on it to have modules as the “dependee” would include the other headers making a mess. We should be fine reverting this change, if indeed is broken. We should be able to propagate the NO_CXXMODULE option through the dependency chain by hand.

**Comment by hahnjo** [comment]
> I am a little confused on how that affects path length.

As I said, I don't understand it either, but the failures are persistent and reproducibly fixed with this change.

> The idea of that patch was that if we specify no_cxxmodule we should not expect targets that depend on it to have modules as the “dependee” would include the other headers making a mess. We should be fine reverting this change, if indeed is broken. We should be able to propagate the NO_CXXMODULE option through the dependency chain by hand.

I understand, but it needs to be implemented differently so that it doesn't omit dependencies to `libCore`.

**Comment by vgvassilev** [comment]
@hahnjo, is there a way to move forward without losing progress we made for windows?

cc: @bellenot

**Comment by hahnjo** [comment]
> @hahnjo, is there a way to move forward without losing progress we made for windows?

I don't see one, please see my very lengthy and detailed commit message of why I think it's not going to work like that.

**Comment by vgvassilev** [comment]
> > @hahnjo, is there a way to move forward without losing progress we made for windows?
> 
> I don't see one, please see my very lengthy and detailed commit message of why I think it's not going to work like that.

Would it not help if we annotate all dependencies by hand all of the dependencies that cannot work with modules?

**Comment by hahnjo** [comment]
> > > @hahnjo, is there a way to move forward without losing progress we made for windows?
> > 
> > I don't see one, please see my very lengthy and detailed commit message of why I think it's not going to work like that.
> 
> Would it not help if we annotate all dependencies by hand all of the dependencies that cannot work with modules?

Yes, this is the direction I've also been thinking about. But it's certainly not an easy change of a couple of lines, and I really would like to fix the currently broken state first. To be clear, this is currently *blocking* me from making changes in ROOT / Cling, or at least making it unnecessarily hard because I always have to either deal with 20-ish unrelated test failures or base my branches on these fixes. It's really painful.

**Comment by vgvassilev** [comment]
Ok, then let's unblock you first.

**Review by vgvassilev — APPROVED** [review]
LGTM, assuming somehow we track proposed alternative of annotating all unsupported dictionaries on Windows..

**Comment by hahnjo** [comment]
> LGTM, assuming somehow we track proposed alternative of annotating all unsupported dictionaries on Windows..

I can open a follow-up issue, would that do the job?

**Pull request merged**